### ISSUE TITLE
Rule proposal: `prefer-double-boolean-negation` [Prefer `!!...` over `Boolean(...)`]

### ISSUE THREAD

**Issue by B2o5T** [typescript-eslint/typescript-eslint#6316]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Prefer `!!...` over `Boolean(...)` since `!!` has the advantage that TypeScript infers a narrow literal boolean `type: true`, while inferring the `Boolean(...)` as `type: boolean`

### Fail Cases

```typescript
var replace = Boolean(foo);
```

### Pass Cases

```typescript
const a = !!foo;
const b = list.filter(Boolean);
```

### Additional Info

https://www.typescriptlang.org/docs/handbook/2/narrowing.html#truthiness-narrowing

Real life case:
https://www.typescriptlang.org/play#code/C4TwDgpgBAsghgYwBYEsB20C8UDeAoKKOALlyNIEYBfAGgKgCNSdHKoq8qoAfXekqADM4AGwDOEOhzyCArmgTAUAezRRkEBAGsAQhEHKAThACCs4MoBiKAB4AKALal4ydBACUfQikFQ7AQn8HADo4T3xCQhCGeg5pOQUlVXUkTS0TQWAIQzMLQVtHZ0RUDHD6Hz8dZWURCDg0R1D3MsioaNjOPCA

**Comment by JoshuaKGoldberg** [comment]
In theory, `!!...` and `Boolean(...)` should be equivalent to the type system. Preferring one over the other is really just a stylistic preference.

In actuality, TypeScript still has https://github.com/Microsoft/TypeScript/16655 open. Which _was_ "fixed" by https://github.com/Microsoft/TypeScript/29955, but had to be reverted: https://github.com/microsoft/TypeScript/issues/16655#issuecomment-494907057. So this is a bug in TypeScript.

In general we try not to add rules that only account for bugs in TypeScript. It'd be better to put effort towards sending a PR to fix the issue in TypeScript itself.

Thanks for filing though!

**Comment by B2o5T** [comment]
@JoshuaKGoldberg theory and practice are different way

in practice, official documentation says
<img width="914" alt="image" src="https://user-images.githubusercontent.com/7361780/211387361-4b7f9b5a-212b-450a-a29c-bebedaf6effd.png">

**Comment by JoshuaKGoldberg** [comment]
Indeed, the official docs are correct. They will need to be updated once someone fixes the issue in TypeScript. 🙂

**Comment by jguddas** [comment]
I could add a `prefer-boolean-call` option here if you like :sweat_smile: 
https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1922/files

**Comment by B2o5T** [comment]
@jguddas Why is this name for the option? Anyway, if some rule could report these mistakes would be nice.

We already catch these cases with no-restrict-syntax

https://github.com/the-guild-org/shared-config/blob/main/packages/eslint-config/src/constants.js#L27-L32

And having auto fix will be awesome 🥳

**Comment by jguddas** [comment]
> @jguddas Why is this name for the option?

I'm terrible at naming variables.

My idea was that `'prefer-boolean-call': true` would replace `!!` with `Boolean(…)` and `'prefer-boolean-call': false` would replace `Bolean(…)` with `!!`.

If you have a better naming idea let me know.

**Comment by bmish** [comment]
For context, here's the ESLint rule for the opposite (`!!foo` -> `Boolean(foo)`): [no-implicit-coercion](https://eslint.org/docs/latest/rules/no-implicit-coercion).
